Reduce duplication in alembic configuration

MatthewWilkes opened the pull request:
Hey @zupo 

After diving more deeply into the approach I had in the other PR, I've decided that it is fundamentally flawed. As is the config directive approach that there's a sketch of in the deferred library that I was planning on extending. I believe that to make those two work we'd have to either monkeypatch alembic or wrap the alembic frontend in our own entrypoint, both of which feel too invasive a change.

This PR represents what I believe is the simpler approach that respects your clear desire for less duplication.

What this does is decouples the alembic and development/production/test ini files, so the alembic configuration is not responsible for database config and the app configuration is not responsible for alembic config.

For users, this means a different invocation of alembic, it gains an `-x ini=etc/development.ini` and the `-c etc/development.ini` becomes `-c etc/alembic.ini`. It would be possible to drop the `-c` entirely if we moved the alembic.ini to the root of the project.

What do you think of this approach?

zupo commented on src/conduit/migrations/env.py:
```diff
@@ -42,10 +43,15 @@ def run_migrations_online():
+    allowed = {"sqlalchemy.url", "sqlalchemy.echo"}
+    settings = {
+        key: value for (key, value) in registry.settings.items() if key in allowed
+    }
```
1. Isn't usage of `get_current_registry` seriously discouraged, especially outside of test suites?
2. Why do we need `allowed`? Security considerations or?

MatthewWilkes replied:
1. The main reason that `get_current_registry` is discouraged, AIUI, it that it's a sign that you've designed your app wrong. If you need the registry you should have access to a request. Here we're trying to hook into the behaviour of a different tool (alembic) when it's run directly and as part of test suite setup. I don't feel bad about `get_current_registry`, myself. If you do, we can use `Base.metadata.bind.url`, but that will prevent the setting `sqlalchemy.echo` working in alembic.
2. It's because the registry's settings can contain other strings that look like sqlalchemy settings, specifically `sqlalchemy.engine`, which is an error to pass to this function.

zupo replied:
1. I've never used `sqlalchemy.echo`, so personally, I'd prefer `Base.metadata.bind.url`. Do you use it or know of people that use it? 
2. Makes sense.
3. One more consideration: I might be completely wrong but I think alembic takes logging configuration from `dev/prod.ini`. Which would mean that there wouldn't be any output after this PR is merged? Let me do some tests.

MatthewWilkes replied:
I do use `sqlalchemy.echo`, but never in alembic, so I think this is fine. I'll make the change to use the Base metadata. It does make for a slightly more readable file, but I don't think we should worry about using `get_current_registry` in this kind of context if it's needed in future.

MatthewWilkes commented:
> (-) Configuration is split, some is in dev/prod.ini some in alembic.ini.
I think this objection is the key one. For me, dev, prod and test are deployment configurations. They declare how the app itself should be spun up. alembic.ini is a tooling configuration.

You've got a `.flake8` file which is an INI format, but that presumably feels natural to have split out from the `development.ini`? I am seeing it the same way, that alembic is a tool that operates on the code, rather than a core part of the application.

It seems to me that the only valid parameter for `alembic.ini` that you'd want to change across deployment types is `sqlalchemy.url`, and that one's brought in from the app config instead. Things like `file_template` and `output_encoding` are pretty orthogonal.

zupo commented:
> You've got a .flake8 file which is an INI format, but that presumably feels natural to have split out from the development.ini? I am seeing it the same way, that alembic is a tool that operates on the code, rather than a core part of the application. Maybe moving alembic.ini to its default position (i.e. not in etc/) would make that clearer?

Very good point!

MatthewWilkes commented:
So, this uses the bind url now, without getting the current registry. Are you still nervous about this split approach, @zupo? I didn't move the `alembic.ini`, but am happy to and update the docs/scripts if that'd help.

zupo commented:
I've gotten used to the split, it makes sense after the initial emotional reaction. 

Yes, please go ahead and move `alembic.ini` to its usual locaiton.

MatthewWilkes commented:
Moved alembic ini and updated the docs and scripts.

Worth noting that as #26 uses alembic in its new fixture, the line https://github.com/niteoweb/pyramid-realworld-example-app/pull/26/files#diff-1293fec17c9fd30efff1a34fe6e8d1f1R23 will need to be updated to be `Config("alembic.ini")` when whichever the latter of #25 and #26 is merged.
